### ISSUE TITLE
Reduce noise/Allow quiet working

### ISSUE THREAD

**Issue by kaste** [SublimeLinter/SublimeLinter#1149]
User (short) stories (from various comments):

> I want to disable everything to write pseudo-code or something quick.

> I open a random file on my disk, or from the internet, which obviously doesn't match my lint style if any at all.

> I just want to *read* code.

Proposed *solutions* are:

- Enable/disable linting (per shortcut) of a view
- Enable/disable SublimeLinter (per shortcut)
- Quickly toggle highlighted regions (with *instant* redraws!).

Related #1101, #98

**Comment by braver** [comment]
Switching off the views seems like a nice and easy way to go about it? Perhaps you don’t care that the linter runs, but you just don’t want to know about it right now.

**Comment by kaste** [comment]
This will most likely be a moderated thread. Please add user stories that vastly differ from what is already in the first description.

**Comment by braver** [comment]
Ok, I can't think of any. At least none that cannot be dealt with using project configuration for instance. I do have opinions on the proposed solutions in relation to the user stories. Want them now or later?

**Comment by kaste** [comment]
I hope we get some more feedback. I only do bugfixes anyway as long as we have real crashes.

I'm especially unsure about: toggle for one view, the current window, or globally.

**Comment by FichteFoll** [comment]
>I'm especially unsure about: toggle for one view, the current window, or globally.

Imo all of these are useful, though personally I found disabling it for the current view to be the most frequently needed version.

**Comment by braver** [comment]
> I'm especially unsure about: toggle for one view, the current window, or globally.

For user story 2, per view wouldn't be enough but per window would. I think per window would be achieved by looping over all views in the window so you kinda get that for free. So, I think per view makes the most sense unless I think of another user story.

**Comment by FichteFoll** [comment]
When implementing per-view settings, you need to acces view settings anyway, so you might as well use the same setting name in project settings and get both functionalities at the same time. The disable state would be saved across sessions for free as well.

I just thought of a really neat UX using Input Handlers, but unfortunately those aren't available in the stable build iirc.

**Comment by Mattiaslndstrm** [comment]
Two thoughts:

**Comment by Mattiaslndstrm** [comment]
Two thoughts: 

For current view would be most useful for me. It would only be confusing if I am to switch file and find that it's not linting. It might also be confusing to remember if I have disabled linting on a file, so maybe keeping one error telling the user that linting is disabled would be a good idea?

Ignore current error would be incredible useful. I can't find any way of doing this. It's quite often I find myself wanting to get rid of an error that I find irrelevant in the current context. But this is a different idea. Should I open a new issue for this?

**Comment by braver** [comment]
> But this is a different idea. Should I open a new issue for this?

No, it’ll just be a won’t fix. Configuring ignore rules is up to the linter to provide. Most, like eslint and flake8, allow you to add comments in the code for situational stuff like that.

But. User stories, people. Focus! ☺️

**Comment by kaste** [comment]
"Ignore current error" is removing noise, so I think that's interesting although I don't know exactly what you mean. What we don't wan't here some sophisticated persisting of these settings or configuring linters. So, basically these are toggles where you usually get to normal mode.

**Comment by gerardroche** [comment]
I solved some of these with a custom command to enable, disable, and toggle linters: 

* https://github.com/gerardroche/sublimefiles/blob/e236b013fbc0051eb1b6af13384dca1efb8cfef4/User/plugin.py#L78

Command Palette: https://github.com/gerardroche/sublimefiles/blob/e236b013fbc0051eb1b6af13384dca1efb8cfef4/User/Default.sublime-commands#L38

```
    { "caption": "SublimeLinter: Disable Linter", "command": "linters", "args": {"action": "disable"} },
    { "caption": "SublimeLinter: Enable Linter", "command": "linters", "args": {"action": "enable"} },
    { "caption": "SublimeLinter: Toggle Linter", "command": "linters", "args": {"action": "toggle"} },
    { "caption": "SublimeLinter: Toggle Mode", "command": "linters", "args": {"action": "toggle_mode"} },
```

Some of my thoughts:

My custom command is just a quick and dirty fix to allow me to enable, disable, toggle, and configure the linters. Some linters are resource intensive and cause my cpu to spike, which is why I've also increased the delay, currently at 0.3, but I'll probably increase it more. 

Doing a search and replace across a lot of files is also a concern and can cause performance issues e.g. opening 50 files with flake8, mypy, means 50x2 lints all fighting for resources. I sometimes disable the linters before doing lots of search and replaces.

I also disable linters if I am writing pseudo-code too. Or even just messy code to get something done. Usually when I'm finished I'll switch the linters on and check the code. Sometimes they just get in the way when you're trying to get something done.

I would say: don't add default key bindings for any of these type of commands. The reasons are that you'll find most people will never use them, fewer will remember them, because these commands are not used all that often. Provide command palette commands, which don't get in anyones way, and provide docs on how to setup custom keybindings with suggested defaults, like this:  https://github.com/gerardroche/sublime-phpunit#key-bindings.

My custom commands are still not ideal, they get the job done. The command that I would next implement if I get the chance would be enable/disable/toggle all linters for current session: I'd like to be able to disable everything for the current session, such that on restart it will reset to whatever the previous state was e.g. if linters are on then disable linters -> restart -> now linters are enabled again. The reason is because sometimes I forget that I disabled a linter.

A quick disable linting for current view is also a good idea. I've ran into the issue of just wanting to read code too.

**Comment by braver** [comment]
Thanks for your thoughts!

> I would say: don't add default key bindings for any of these type of commands

We will definitely not introduce new key bindings. Packages shouldn't do that in the first place (there are only so many keys on the keyboard after all), so we're sticking with what we've had for years.

**Comment by humatic** [comment]
Use case: I like to only run the linter when I'm ready to commit my code. Otherwise I don't want to see the highlighting popping in and out of existence while I write. Also sometimes I intentionally ignore style rules during prototyping, for expedience.

+1 Toggle per-view especially. To mitigate some of the potential confusion about whether linting is on or off, could there be a short string in the bottom status bar next to tab and syntax info? For example: "Linter: flake8   Spaces: 4   Python"

**Comment by braver** [comment]
> could there be a short string in the bottom status bar next to tab and syntax info? 

Good idea, could you put that in  separate issue?

**Comment by kaste** [comment]
For status bar info use #1205 

For the next release we will have a quick highlights toggle via #1192 

For the next release: @gerardroche To enable/disable linters. `persist.linter_classes` is a `Dict[linter_name, linter_class]`. To disable a linter you basically set `linter_class.disabled = True`, after that you fire `sublime.run_command('sublime_linter_config_changed')`.

**Comment by kaste** [comment]
Next release == 4.1

**Comment by Relequestual** [comment]
Looks like this has been implemented, released, and works. Can probably close?

**Comment by humatic** [comment]
Now it works great for me! I just bound the toggle to the F8 key to turn on/off when I want.
```python
{ "keys": ["f8"], "command": "sublime_linter_toggle_highlights" },
```

**Comment by braver** [comment]
> Looks like this has been implemented, released, and works. Can probably close?

Well, we can probably do more if there is a need for it, but the toggle seems to cover a lot of cases already.

**Comment by kaste** [comment]
Here is a linter toggler plugin https://github.com/kaste/SublimeLinter-addon-toggler

Note: it's a global toggle, not per view :-( )

**Comment by kaste** [comment]
Sorry, the toggler plugin will only work with a next release or master. Thought we shipped already.

**Comment by kaste** [comment]
So, we shipped 4.3 and my toggle addon should work. We still need to decide if we need/want some settings per view. But we now have a dope status bar.

**Comment by braver** [comment]
@kaste I'm 100% ok with keeping the toggler as a separate add-on. If that works, we've seemingly covered all the scenarios of this discussion and we need new input/feedback if we want to improve anything else here.
